### PR TITLE
[BC-BREAKING] Torch.flatten() returns a 1-dim tensor on a 0-dim tensor

### LINKED ISSUE #22963: torch.flatten() returns a 0-dim (not 1-dim) tensor for 0-dim tensors
## 🐛 Bug

torch.flatten works in a counterintuitive way when supplied with 0-dim tensors.

In numpy:

>>> np.asfarray(123).flatten()
array([123.]) # returns a 1-dim array
>>> np.asfarray([1,2,3]).flatten()
array([1., 2., 3.]) # returns a 1-dim array

So, in numpy, flatten() always returns a 1-dim array, which is exactly why one would use it.
In contrast, in pytorch, it returns a 0-dim tensor for 0-dim tensors, which defeats the whole purpose of flatten: to convert all tensors to 1-dim, so we can handle arbitrarily shaped tensors in a uniform way.

In torch:

>>> torch.tensor(123).flatten()
tensor(123) # returns a 0-dim tensor!
>>> torch.tensor([1,2,3]).flatten()
tensor([1, 2, 3]) # returns a 1-dim tensor

## To Reproduce

1. Call .flatten() or torch.flatten() on a 0-dim tensor

## Expected behavior

The tensor becomes 1-dimensional.

## Environment

PyTorch version: 1.2.0a0+c9f41e9
Is debug build: No
CUDA used to build PyTorch: 10.1.105

OS: Ubuntu 16.04.3 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.14.5

Python version: 3.5
Is CUDA available: Yes
CUDA runtime version: 10.1.105
GPU models and configuration: 
GPU 0: TITAN X (Pascal)
GPU 1: TITAN X (Pascal)

Nvidia driver version: 418.56
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.6.0.21
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a

Versions of relevant libraries:
[pip3] numpy (1.16.4)
[pip3] torch (1.2.0a0+c9f41e9)
[pip3] torchfile (0.1.0)
[pip3] torchvision (0.1.9)
[conda] Could not collect

cc @ezyang @SsnL

### PR DESCRIPTION
**Pull request by CamiWilliams**
PR for `torch.flatten()` to return a 1-dim tensor on a 0-dim tensor

> torch.tensor(123).shape -> torch.Size([])
> torch.tensor(123).flatten() -> torch.tensor([123])
> torch.tensor(123).flatten().shape -> torch.Size([1])

resolve #22963

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
There might be some tests that fail after this change so you'll have to check the CI.

**Comment by gchanan** [comment]
nice!
